Plugins/GitUIPluginInterfaces/ObjectId.cs

```diff
@@ -244,19 +326,31 @@ private static bool IsValidCharacters(string s)
         private readonly ulong _i1;
         private readonly ulong _i2;
         private readonly uint _i3;
+        private readonly ulong _i4;
```

Comment:
ObjectId is critical for performance. Even more so when sha256 is used I assume, to be enabled for big repos.
sha1 is stored as 16+16+8 (which was maybe 150ms faster than 5*16 opening 100K commits in Linux if I remember correctly).
sha256 probably would benefit from 4*16 rather than 16+16+8+16+8. (Is it easier for the compiler to optimize 4*16 too?)
We have optimized individual if statements to optimize loading, this change could me measurable (as well as changes in RevisionReader).

It seems like a repo can contain both sha1/sha256 hashes. Maybe a sha256 can be handled as 40 bit.But upcoming Git has changes in the upcoming release:
https://www.phoronix.com/news/Git-2.45-rc0-Released

So maybe not a quick merge...

Reply:
I wonder if we use subtypes here for storage. i.e. make `ObjectId` abstract and have different implementations `Sha1Id`, `Sha256Id` etc.

If we ever need a third implementation, one can imagine it making the approach taken here even more challenging to follow.

Reply:
I will look into subtypes and performance benchmarks for different implementations.

> It seems like a repo can contain both sha1/sha256 hashes. Maybe a sha256 can be handled as 40 bit.But upcoming Git has changes in the upcoming release: https://www.phoronix.com/news/Git-2.45-rc0-Released
> 
> So maybe not a quick merge...

If I understand https://git-scm.com/docs/hash-function-transition#_object_names_on_the_command_line correctly, then the hash written to the console (sha1/sha256) will always be the same during a mode of operation. It's not like some commits are sha1 and some are sha256 in one `git log` output:

> ("dark launch") Treat object names input by the user as SHA-1 and **convert any object names written to output to SHA-1**, but store objects using SHA-256. This allows users to test the code with no visible behavior change except for performance. This allows running even tests that assume the SHA-1 hash function, to sanity-check the behavior of the new mode.
> 
> ("early transition") Allow both SHA-1 and SHA-256 object names in input. **Any object names written to output use SHA-1**. This allows users to continue to make use of SHA-1 to communicate with peers (e.g. by email) that have not migrated yet and prepares for mode 3.
> 
> ("late transition") Allow both SHA-1 and SHA-256 object names in input. **Any object names written to output use SHA-256**. In this mode, users are using a more secure object naming method by default. The disruption is minimal as long as most of their peers are in mode 2 or mode 3.
> 
>("post-transition") Treat object names input by the user as SHA-256 and **write output using SHA-256**. This is safer than mode 3 because there is less risk that input is incorrectly interpreted using the wrong hash function.

If I understand it correctly, `IGitModule` would be the right place to call `git rev-parse --show-object-format=output` once for a repository in order to get the length of all hashes of the repository?

Reply:
Yes, the `GitModule` ctor is the right place.